### PR TITLE
Modify language assertion in AssemblyStack::optimize

### PR DESCRIPTION
**Pull request by bshastry**
### Description

Currently, `AssemblyStack::optimize()` is targeted at `StrictAssembly` only. It was found that one YUL input was somehow successfully parsed, and only at the time of optimization it was found that the following assertion inside `libyul/AsmAnalysis.cpp` would fail

```
        solAssert(m_dialect->flavour != AsmFlavour::Yul, "");
```

The fix in this PR ensures that such failures are caught much earlier inside the call to `AssemblyStack::optimize()`

In addition to modifying the assertion, cmdline test `--yul --optimize` is disabled. This is because the patch rightly fails this test.

### CODE REVIEW

**Review by chriseth — CHANGES_REQUESTED** [review]
This still needs a change in CommandlineInterface to disable the combination of `--yul` and `--optimize`. It should be a simple change around line 839. Please also add a check that these options now fail (similar to the check in cmdlineTests.sh line 334

File: libsolidity/interface/AssemblyStack.cpp [hunk 1]
```diff
@@ -84,7 +84,7 @@ bool AssemblyStack::parseAndAnalyze(std::string const& _sourceName, std::string
 
 void AssemblyStack::optimize()
 {
-	solAssert(m_language != Language::Assembly, "Optimization requested for loose assembly.");
+	solAssert(m_language == Language::StrictAssembly, "Optimization requested for either loose assembly or YUL.");
```
Comment (axic) [inline libsolidity/interface/AssemblyStack.cpp]: Is this change required because the optimiser suite doesn't work with yul? If so, I'd suggest to also add an assertion to `yul::OptimiserSuite::run`

Comment (axic) [reply libsolidity/interface/AssemblyStack.cpp]: And ideally this should be changed from assertion to `solUnimplemented` (see how `assemble` works).

Comment (chriseth) [reply libsolidity/interface/AssemblyStack.cpp]: There are some optimizer stages that introduce EVM opcodes (`eq` for example). Some other stages do not take yul types into account properly. My suggestion would be to fix that once we have a wasm backend prototype.

Comment (bshastry) [reply libsolidity/interface/AssemblyStack.cpp]: > Is this change required because the optimiser suite doesn't work with yul?

My understanding is that the optimizer suite for yul is not fully implemented yet, so we run into known errors if we try to fuzz it.

Comment (bshastry) [reply libsolidity/interface/AssemblyStack.cpp]: @axic I agree, going to change `solAssert` to `solUnimplemented`

**Comment by bshastry** [comment]
> This still needs a change in CommandlineInterface to disable the combination of `--yul` and `--optimize`. It should be a simple change around line 839. Please also add a check that these options now fail (similar to the check in cmdlineTests.sh line 334

Done.

File: solc/CommandLineInterface.cpp [hunk 2]
```diff
@@ -836,13 +836,11 @@ bool CommandLineInterface::processInput()
 				return false;
 			}
 		}
-		if (optimize && inputLanguage == Input::Assembly)
+		if (optimize && (inputLanguage == Input::Assembly || inputLanguage == Input::Yul))
```
Comment (axic) [inline solc/CommandLineInterface.cpp]: Why not `!= Input::StrictAssembly`?

Comment (bshastry) [reply solc/CommandLineInterface.cpp]: Agree :)

**Review by erak — DISMISSED** [review]

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -856,13 +856,11 @@ bool CommandLineInterface::processInput()
 				return false;
 			}
 		}
-		if (optimize && inputLanguage == Input::Assembly)
+		if (optimize && inputLanguage != Input::StrictAssembly)
 		{
 			serr() <<
-				"Optimizer cannot be used for loose assembly. Use --" <<
+				"Optimizer cannot be used for both loose assembly and Yul. Use --" <<
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: This might be interpreted as if the combination of loose assembly and yul cannot be used. What about "Can only be used for strict assembly?"

File: test/cmdlineTests.sh [hunk 4]
```diff
@@ -337,7 +337,6 @@ printTask "Testing assemble, yul, strict-assembly and optimize..."
     # Non-empty code results in non-empty binary representation with optimizations turned off,
     # while it results in empty binary representation with optimizations turned on.
     test_solc_assembly_output "{ let x:u256 := 0:u256 }" "{ let x:u256 := 0:u256 }" "--yul"
-    test_solc_assembly_output "{ let x:u256 := 0:u256 }" "{ }" "--yul --optimize"
```
Comment (chriseth) [inline test/cmdlineTests.sh:340]: Could add a test that checks that it fails.

File: solc/CommandLineInterface.cpp (hunk 3 above)
Comment (bshastry) [reply solc/CommandLineInterface.cpp]: fixed, yes the proposed change is more elegant

File: test/cmdlineTests.sh (hunk 4 above)
Comment (bshastry) [reply test/cmdlineTests.sh:340]: done.

**Review by chriseth — APPROVED** [review]
